Added tabbed forms

nightmarebadger opened the pull request:
Added option to add tabs to forms, using colander.Mapping() (each Mapping appears in a new tab on the form). 

Form validation is changed so all forms with errors are highlighted, and the first tab with errors is automatically selected. Also fixed some problems with radio widgets using integers as ID's, fixing two problems in selenium tests on the way.

Added basic info about adding tabbed forms into the README.

nightmarebadger committed:
Possible to have tabs in forms using deform.Mapping().

nightmarebadger committed:
Added basic tests, added support for forms without a csrf
token.

nightmarebadger committed:
Only colander.Mapping() create new tabs, other types with
children do not anymore.

nightmarebadger committed:
Updated readme with instructions for running tests.

nightmarebadger committed:
Updated README.rst

nightmarebadger committed:
Updated README again as git markdown is messing around with my code o.O

nightmarebadger committed:
Read previous commit ...

nightmarebadger committed:
As before

nightmarebadger committed:
Fixed templates so radio choice works well now

nightmarebadger committed:
Form now jumps to the first tab with validation problems, all
tabs with validation problems are highlighted

nightmarebadger committed:
Fixing tests to work with .typ

nightmarebadger committed:
Updated README with data on tabbed forms.

dnouri commented on README.rst:
```diff
@@ -45,6 +45,31 @@ looks in practice you can run these commands, assuming that you have a
+
+To get a form with the tab *Default*, which has a *number* field, and the tab
+*Mapping*, which has a *name* and *date* field, use the following example. If you
+want to change the title of the tab, use `Mapping(title="New title")`
```
This seems like an unusual example.  It's cool that we fall back to "Default" and all, but it's rather over-represented here with this example, IMO.  What about we just make two named tabs?  We could also use a more classical example like Person, Car, etc.

dnouri commented on deform_bootstrap/templates/form.pt:
```diff
@@ -8,51 +8,89 @@
-  tal:attributes="autocomplete autocomplete">
-  
-  <fieldset>
+              tabifyForm import: deform_bootstrap.utils.tabifyForm;
```
`deform_bootstrap` lowercase, underscore for function names: so `tabify_form`

dnouri commented on README.rst:
```diff
@@ -45,6 +45,31 @@ looks in practice you can run these commands, assuming that you have a
+Mappings in form appear in new tabs. The mapping title is the name of the tab,
+and the data in the mapping appears inside the tab.
+Everything that isn't in a mapping appears inside a '*Default*' tab. If no
+mappings are present, there are no tabs on the form.
```
Given that "Default" does not exist in deform or deform_bootstrap yet, and thus it's not translated, it's a poor choice of words.  Let's maybe use "Basic" instead.  Because "Default" is purely the programmer's view on what this is, it doesn't make any sense to users.

nightmarebadger committed:
Changed according to comments on pull request & extras:
Changed 'Default' tab to 'Basic', updated README with
another example, changed function name to Kotti standards
, field with the first error on forms in now focused.

disko commented on deform_bootstrap/static/deform_bootstrap.css:
```diff
@@ -1121,6 +1121,7 @@ input[type="checkbox"][readonly] {
   border-color: #c09853;
 }
 
+ul#form-tabs li.error a,
```
Can we please make this a class selector instead of an id selector, i.e. `ul.form-tabs`?  Seems more appropriate to me…

disko commented on deform_bootstrap/templates/form.pt:
```diff
@@ -8,51 +8,91 @@
-        </button>
+  <!-- Make the tabs -->
+  <tal:block tal:condition="not: fields_list.only_one">
+    <ul id="form-tabs" class="nav nav-tabs">
```
See above: should be `<ul class="nav nav-tabs form-tabs">` IMO.

disko commented on deform_bootstrap/templates/form.pt:
```diff
@@ -8,51 +8,91 @@
+
+
+  <!-- Populate the tabs -->
+  <div id="form-tabs-content" class="tab-content">
```
Why is the id needed at all here?  Can't we just use the `tab-content` class as a selector?

disko commented on deform_bootstrap/templates/form.pt:
```diff
@@ -71,5 +111,31 @@
+        var alreadySelected = false;
+
+        // If we get some errors, remove the active classes
+        if($("div#form-tabs-content fieldset div.error").length !== 0){
```
These had to be changed too, if you agree on my previous comments.

nightmarebadger committed:
Got rid of some ID's in favor of classes

kstrauser commented:
What would be a good approach for making this configurable? We don't want to use tabs for each mapping and were surprised to see them popping up all over our UI. I was thinking of an approach like setting `tabbed_mappings = True` on a colander.Schema to enable this behavior, but what would be the right place for storing that configuration information?

dairiki commented:
FWIW, I agree that the tabbed forms should be configurable.  And they should be disabled by default, since it does result in a shocking change of appearance.

I think the option (if one is really needed) belongs on `deform.widget.FormWidget`.  Really a new option is not needed — one can just point `FormWidget.template` to a tabbed (or non-tabbed) version of the form template.

dnouri commented:
Thanks for your comments.  It looks like we made a wrong decision when we decided to make the tabbed forms the unconfigurable default.

I agree with @dairiki's suggestion of setting `FormWidget.template` explicitly when the tabbed forms appearance is desired.

cc @nightmarebadger

kstrauser commented:
@dnouri Just to be clear, I really like the tabbed forms and think they're a great option when appropriate. We were just surprised at the fairly large change in a patch release.

dairiki commented:
Apologies if I sounded a little harsh.  I totally agree with @kstrauser that the tabbed layout is a great option to have.

dnouri commented:
No offense taken.  Your feedback was most constructive.

haitaoli-eng commented:
Hi guys,
I really love the tabbed form and thanks for providing this great function.

Recently I've been trying to internationalize my webapp by following the deform doc (http://deform.readthedocs.org/en/latest/i18n.html). However, after changing from the default renderer to the customised ZPT renderer where you can see in the link, I lose the tabbed forms.  All the contents are now displayed in a vertical manner. Any hints on how to get this problem fixed?

Any help will be greatly appreciated!

Cheers,
James

haitaoli-eng commented:
@dnouri @disko

disko commented:
@J20S Tabbed forms don't exist in Deform, they're a feature of deform_bootstrap.  You need to add the deform_bootstrap template dir to your custom renderer.  

Please also note, that deform_bootstrap is no longer maintained.  It has basically been merged into deform 2, without the tabs feature though (IIRC).

haitaoli-eng commented:
@disko Thanks for your comment and it worked!!
Just so that everyone else knows, if you want to internationalize your webapp and do not want to lose the tabbed forms supported by bootstrap_deform. Make sure you create a customised ZPT renderer in the following way:

deform_template_dir = resource_filename('deform', 'templates/')
deform_boostrap_template_dir = resource_filename('deform_bootstrap', 'templates/')
def translator(term):
    return get_localizer(get_current_request()).translate(term)
zpt_renderer = deform.ZPTRendererFactory(
    [deform_boostrap_template_dir,deform_template_dir],
    translator=translator)
form = deform.Form(schema, buttons=(_('Submit'),), renderer=zpt_renderer)

Please note, the order of the template directories does matter. Make sure deform_bootstrap template directory is before the deform template directory.
